docs/proposals/event-notification.md

```diff
@@ -0,0 +1,141 @@
+# ODO build and application status notification 
+
+## Summary
+
+With this proposal and [it's related issue](https://github.com/openshift/odo/issues/2550) we examine how to consume build/run event output from odo in order to allow external tools (such as IDEs) to determine the application/build status of an odo-managed application.
+
+In short, the proposal is:
+- New flag for push command, `odo push -o json`: Outputs JSON events that correspond to what devfile actions/commands are being executed by push.
+- New odo command `odo component status -o json`: Pings the application URLs, and checks the container/pod status every X seconds (and/or a watch, for Kubernetes case). The result of both is output as JSON to be used by external tools to determine if the application is running.
+- A standardized markup format for communicating detailed build status, something like: `#devfile-status# {"buildStatus":"Compiling application"}`, to be optionally included in devfile scripts that wish to provide detailed build status to consuming tools.
+
+## Terminology
+
+With this issue, we are looking at adding odo support for allowing external tools to gather build status and application status. We further divide both statuses into detailed and non-detailed, with detailed principally being related to statuses that can be determine by looking at container logs.
+
+**Build status**: A simple build status indicating whether the build is running (y/n), and whether the last build succeeded or failed. This can be determined based on whether odo is running a dev file action/command, and the process error code (0 = success, non-0 = failed) of that action/command.
+
+**Detailed build status**: An indication of which step in the build process the build is in. For example: are we 'Compiling application', or are we 'Running unit tests'?
+
+**App status**: Determine if an application is running, using container status (for both local and Kubernetes, various status: containercreating, started, restarting, etc), or whether an HTTP/S request to the root application URL returns an HTTP response with any status code.
```

Comment:
Due to the use of the SupervisorD the application might not be running or falling and the container would be still running. It needs to check the application process to determine if the application is running or not.

Reply:
Agreed: the use of container status here is an imperfect representation of the app status for that exact reason. It's still a useful data point (eg if the container ISN'T running. then the app is obviously not running), but if the container IS running it doesn't fully/necessarily imply the app is running (due to SupervisorD).

I like the idea of checking the application process itself, let me check the feasibility of that...

Reply:
K8s health checks appear not to be defined in either odo devfile v1 or v2 (correct me if I'm wrong, I didn't see it), so this would be something to consider if/when these are added there. (And these would not apply if odo was running with docker target)

Reply:
Not a bad idea, but I'm afraid that using k8s health checks might have unwanted consequences. I wouldn't want to use LivelinessProbe as once it starts failing Kubernetes will restart the Pod. Using ReadinessProbe might work, but if ReadinessProbe starts to fail it Kubernetes will stop sending traffic to the Service, It might not be what you want when developing an application.

Reply:
@kadel that leaves us with startup probes which, from the documentation, sound like what might fit in. However, it would only make sense to use those if integrating those won't be a problem. Otherwise, coming up with our own solution for that shouldn't be a hassle, I guess.

Reply:
A followup: as discussed earlier in the thread, `supervisord ctl status` can be used to verify that long-running processes inside the container are running as expected.

I have prototyped it and it works well, so I've updated the proposal to include these data as another event type produced by `component status`.